aspnetcore/fundamentals/primitives/change-tokens.md

```diff
@@ -0,0 +1,192 @@
+* Sets the `CurrentState` property string to a descriptive message that records the time that the code ran.
+* Notes the current `state` in its `WriteConsole` output.
+
+[!code-csharp[Main](change-tokens/sample/Extensions/ConfigurationMonitor.cs?name=snippet3)]
+
+An instance `ConfigurationMonitor` is registered as a service in `ConfigureServices` of *Startup.cs*:
+
+[!code-csharp[Main](change-tokens/sample/Startup.cs?name=snippet1)]
+
+The Index page offers the user control over configuration monitoring. The instance of `IConfigurationMonitor` is injected into the `IndexModel`:
+
+[!code-csharp[Main](change-tokens/sample/Pages/Index.cshtml.cs?name=snippet1)]
+
+The app responds on form submits when the user selects buttons in the interface to enable and disable monitoring:
+
+[!code-cshtml[Main](change-tokens/sample/Pages/Index.cshtml?range=35)]
+
+[!code-csharp[Main](change-tokens/sample/Pages/Index.cshtml.cs?name=snippet2)]
+
+When the user triggers the `OnPostStartMonitoring` method, monitoring is enabled and the current state is cleared. When the user triggers the `OnPostStopMonitoring` method, monitoring is disabled and the state is set to reflect that monitoring is not occurring.
+
+## Monitoring cached file changes
+
+File content can be cached in-memory using [IMemoryCache](/dotnet/api/microsoft.extensions.caching.memory.imemorycache). In-memory caching is described in the [In-memory caching](xref:performance/caching/memory) topic. Without taking additional steps, *stale* (outdated) data is returned from a cache if the source data has changed.
+
+For example, ignoring the status of a source file for cached file data on a [sliding expiration](/dotnet/api/microsoft.extensions.caching.memory.memorycacheentryoptions.slidingexpiration) period leads to stale cache data when the file is changed but the caching system isn't notified. Each request for the data renews the sliding expiration period, but the file is never reloaded into the cache. Any app features that use the file's cached content are subject to possibly receiving stale content.
+
+Using change tokens in a file caching scenario avoids the possibility of getting stale file content from the cache. The sample app demonstrates an implementation of the approach.
+
+The sample establishes a static method, `GetFileContent`, to return file content. The method uses a retry algorithm with exponential back-off to cover cases where a file lock is temporarily preventing a file from being read (*Utilities/Utilities.cs*):
+
+[!code-csharp[Main](change-tokens/sample/Utilities/Utilities.cs?name=snippet2)]
+
+A `FileService` is created to handle cached file lookups. The `GetFileContent` method call of the service attempts to obtain file content from the in-memory cache and return it to the caller (*Services/FileService.cs*).
+
+If cached content isn't found using the cache key, which is the file path in this example, the following actions are taken:
```

Comment:
~which is the file path in this example~
Avoid the temptation to provide too many details.